content/docs/user-guide/dvc-files.md

```diff
@@ -240,14 +248,245 @@ stages:
+templated/parameterized `dvc.yaml` files (see next section), their actual values
+are substituted into the `dvc.lock` YAML structure.
+
+### Parameterize `dvc.yaml`
+
+`dvc.yaml` supports a templating format to reuse values from different sources
+in the YAML structure itself. The sources can be external
+[parameters files](/doc/command-reference/params), or internal `vars`, as
+detailed below.
+
+Let's say we have `params.yaml` (default params file) with the following
+contents:
+
+```yaml
+models:
+  us:
+    threshold: 10
+    filename: 'model-us.hdf5'
+```
+
+Those values can be used anywhere in `dvc.yaml` with the `${}` _substitution
+expression_:
+
+<!-- prettier-ignore-start -->
+```yaml
+stages:
+  build-us:
+    cmd: >-
+      python train.py
+      --thresh ${models.us.threshold}
+      --out ${models.us.filename}
+    outs:
+      - ${models.us.filename}:
+        cache: true
+```
+<!-- prettier-ignore-end -->
+
+DVC will track simple param values (numbers, strings, etc.) used in `${}` (they
+will be listed by `dvc params diff`).
+
+Alternatively, values for substitution can be listed as top-level `vars` like
+this:
+
+```yaml
+vars:
+  - models:
+      us:
+        threshold: 10
+  - desc: 'Reusable description'
+
+stages:
+  build-us:
+    desc: ${desc}
+    cmd: python train.py --thresh ${models.us.threshold}
+```
+
+> DVC merges external params and `vars` as long as there are no leaf node
+> collisions (so the two examples above can't be used simultaneously). For
+> example, `{"grp": {"a": 1}}` can be merged with `{"grp": {"b": 2}}`, but not
+> with `{"grp": {"a": 7}}`.
+
+⚠️ Note that doesn't track values from `vars` like parameters (ignored by
+`dvc params diff`).
+
+To load additional params files, list them in the top `vars`, in the desired
+order, e.g.:
+
+> Their paths will be evaluated based on `wdir`, if one given.
+
+```yaml
+vars:
+  - params.json
+  - myvar: 'value'
+  - config/myapp.yaml
+```
+
+(ℹ️) Note that the default `params.yaml` file is always loaded first, if
+present.
+
+It's also possible to specify what to include from additional params files, with
+a `:` colon:
+
+```yaml
+vars:
+  - params.json:clean,feats
+
+stages:
+  featurize:
+    cmd: ${feats.exec}
+    deps:
+      - ${clean.filename}
+    outs:
+      - ${feats.dirname}
+```
+
+Stage-specific values are also supported, with inner `vars`. You may also load
+additional params files locally. For example:
+
+```yaml
+stages:
+  build-us:
+    vars:
+      - params.json:build
+      - model:
+        filename: 'model-us.hdf5'
+    cmd: python train.py ${build.epochs} --out ${model.filename}
+    outs:
+      - ${model.filename}
+```
+
+⚠️ Important: Limitations of local `vars`:
+
+- `wdir` cannot use values from local `vars`, as DVC uses the working directory
+  first (to load any values from params files listed in `vars`).
+- `foreach` is also incompatible with local `vars` at the moment.
+
+The substitution expression supports these forms:
```

Comment:
just to check, how correct the `substitution expression` term is?

Reply:
it also feels that we should define this way before we jumped into different options for `vars`

Reply:
> how correct the substitution expression term is?

Will research ⌛ 

> should define this way before

We do intro the term early on before using `${}` in any example. Do you mean putting "substitution expression supports these forms..." before dvc.yaml examples?

Reply:
OK @shcheklein, on the question of the "substitution expression" I here some the alternatives:

- We call it "`${}` expression" and use "interpolation" to explain it in the original [wiki page](https://github.com/iterative/dvc/wiki/Parametrization), but that latter term seemed too mathematical to me.
- [Helm](https://helm.sh/docs/chart_template_guide/getting_started/) (K8 YAML templating) calls it a "directive" (e.g. `{{ .Something }}`) and uses words "value", "generated", and "render/inject/insert" for the explanations.
  > BTW apparently there are some [strong opinions](http://leebriggs.co.uk/blog/2019/02/07/why-are-we-templating-yaml.html) out there against templating YAML 😆
- [Kustomize](https://kubectl.docs.kubernetes.io/guides/introduction/kustomize/) (YAML) uses a different approach (template-free) with terms like "generate", "variants" and "overlays" — not directly comparable but listing here because it's an interesting alternative, maybe we can consider adopting something like that in the future.
- [Handlebars](https://handlebarsjs.com/guide/) (HTML/JS) calls them "handlebars expressions" (e.g. `{{prop}}`); uses "generate/render", "property/value", "replace"
- [Twig](https://twig.symfony.com/doc/3.x/api.html)/[Django](https://docs.djangoproject.com/en/3.1/topics/templates/) (HTML/PHP): "tags/filters" (e.g. `{% var %}`, `{{ function }}`); "render", "generate", "values"
- [Mustache](https://mustache.github.io/mustache.5.html) (Ruby, JS, etc.): "tags" (e.g. `{{key}}`); "render", "replace", "values".
- [Go](https://golang.org/pkg/text/template/) (text): "actions" (e.g. `{{ "string" }}`); "values", "evaluation", "generate"
- [Bash](https://www.gnu.org/software/bash/manual/html_node/Command-Substitution.html) calls `$(command)` a "command substitution"

Conclusions:
- "Substitution" is not very used out there, I guess the `$` in our syntax reminded me more of Bash than of modern templating engines.
- For `${}`, just "expression" is OK. Most templating formats have a special term though, usually "tag" for value injections. 
- For related terms, I think in our case "replace" and "generate" are the most appropriate.